research/minigo/dual_net.py

```diff
@@ -0,0 +1,209 @@
+# Copyright 2018 MLBenchmark Group. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+"""Contains utility and supporting functions for DualNet.
+
+This module provides the model interface, including functions for DualNet model
+bootstrap, training, validation, loading and exporting.
+"""
+
+# pylint: disable=g-bad-import-order
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import os
+
+import tensorflow as tf
+
+import dualnet_model
+import features
+import go
+import model_utils
+import preprocessing
+import symmetries
+
+
+class DualNetwork(object):
+  """The DualNetwork class for the complete model with graph and weights.
+
+  This class can restore the model from saved files, and provide inference for
+  given examples.
+  """
+
+  def __init__(self, save_file, params):
+    """Initialize the dual network from saved model/checkpoints.
+
+    Args:
+      save_file: Path where model parameters were previously saved.
+      params: Hyper parameters for the dual network
+    """
+    self.save_file = save_file
+    self.hparams = params
+    self.inference_input = None
+    self.inference_output = None
+    config = tf.ConfigProto()
+    config.gpu_options.allow_growth = True
```

Comment:
This is interesting. Why do we set this here? That changes the behavior WRT GPU mem usage. @bitfort -- what does the addition/removal of this change? My understanding is that this tries to limit gpu mem, but has some strange effects in certain cases.

Reply:
I checked this link for using GPUs: https://www.tensorflow.org/programmers_guide/using_gpu. From this post, I think to set the allow_growth to True is trying to allow the GPU memory growth during the process execution, rather than pre-mapping all GPU memory to all GPUs at the begging. Not sure if this is correct. @bitfort Could you clarify on this?